File: packages/js/core/src/algorithms/filter-results.ts
import { Tracer } from "@polywrap/tracing-js";

/* eslint-disable @typescript-eslint/no-explicit-any */
export const filterResults = Tracer.traceFunc(
  "core: filterResults",
  (result: unknown, filter: Record<string, any>): unknown => {
    if (!result) {
      return result;
    }

    if (typeof result !== "object") {
      throw Error(
        `The result given is not an object. ` +
          `Filters can only be given on results that are of 'object' type.\n` +
          `Filter: ${JSON.stringify(filter, null, 2)}`
      );
    }

    const filtered: Record<string, any> = {};
    const res: any = result;

    for (const key of Object.keys(filter)) {
      if (res[key] !== undefined) {
        if (typeof filter[key] === "boolean") {
          filtered[key] = res[key];
        } else {
          filtered[key] = filterResults(res[key], filter[key]);
        }
      } else {
        filtered[key] = undefined;
      }
    }

    return filtered;
  }
);

File: packages/py/core/src/algorithms/filter_results.py
from json import dumps
from typing import Any, Dict


def filter_results(result: Any, filter: Dict[str, Any]) -> Any:
    if not result:
        return result

    if not isinstance(result, dict):
        raise ValueError(
            f"""The result given is not a dict. 
            Filters can only be given on results that are of 'dict' type.\n
            Filter: ${dumps(filter)}"""
        )

    filtered = {}
    for key in filter:
        if result[key] is not None:
            if isinstance(filter[key], bool):
                filtered[key] = result[key]
            else:
                filtered[key] = filter_results(result[key], filter[key])
        else:
            filtered[key] = None

    return filtered
